Bugfix:  deadline checking existing frames fails when there is number in file name

kalisp opened the pull request:
## Changelog Description
Previous implementation of validator failed on files with any other number in rendered file names.

Used regular expression pattern now handles numbers in the file names  (eg "Main_beauty.v001.1001.exr", "Main_beauty_v001.1001.exr", "Main_beauty.1001.1001.exr") but not numbers behind frames (eg. "Main_beauty.1001.v001.exr")


## Testing notes:
1. create render instance with some number in it name (you could use Variant for that)
2. publish this through Deadline (or Unreal)

iLLiCiTiT commented on openpype/plugins/publish/validate_sequence_frames.py:
```diff
@@ -7,28 +12,45 @@ class ValidateSequenceFrames(pyblish.api.InstancePlugin):
-
-        missing = collection.holes().indexes
-        assert not missing, "Missing frames: %s" % (missing,)
+        representations = instance.data.get("representations")
```
```suggestion
        representations = instance.data.get("representations")
        if not representations:
            return
```

kalisp commented on openpype/plugins/publish/validate_sequence_frames.py:
```diff
@@ -7,28 +12,45 @@ class ValidateSequenceFrames(pyblish.api.InstancePlugin):
-    families = ["imagesequence"]
-    hosts = ["shell"]
+    families = ["imagesequence", "render"]
+    hosts = ["shell", "unreal"]
```
For Unreal it is happening locally.

BigRoy commented on tests/unit/openpype/plugins/publish/test_validate_sequence_frames.py:
```diff
@@ -0,0 +1,144 @@
+from pyblish.api import Instance as PyblishInstance
+
+from tests.lib.testing_classes import BaseTest
+from openpype.plugins.publish.validate_sequence_frames import (
```
Are we sure this works, the folder doesn't have an `__init__.py` right?

kalisp replied:
It is working for me in Pycharm, wanted to tested directly from command line, but didn't get to it.

kalisp replied:
Works for me even from command line with `start.py`

iLLiCiTiT replied:
if you specify full "path" with dots, subfolders don't have to contain `__init__.py`. (`import openpype.plugins.publish.validate_sequence_frames` would work but `from openpype.plugins.publish import validate_sequence_frames` would not)

BigRoy commented on openpype/plugins/publish/validate_sequence_frames.py:
```diff
@@ -7,28 +12,47 @@ class ValidateSequenceFrames(pyblish.api.InstancePlugin):
+            if isinstance(repr_files, str):
+                repr_files = [repr_files]
+
+            _, ext = os.path.splitext(repr_files[0])
+            pattern = r"\D?(?P<index>(?P<padding>0*)\d+){}$".format(
+                re.escape(ext))
```
Side note: It's good to be aware that this logic will fail on any extensions that include a dot, like e.g. `ass.gz` unless we make sure `ext` value includes the full extension which is what `splitext` excludes.

kalisp replied:
Would it be better to depend on `ext` field in representation? Would there be `ass.gz`?

BigRoy replied:
Arnold can export ass files gzipped from Houdini, the extension is `.ass.gz` yes.

kalisp replied:
Ext on representation is used if present, if not ext is parsed out.

BigRoy commented on tests/unit/openpype/plugins/publish/test_validate_sequence_frames.py:
Can we also add tests to ensure whatever logic we create will also match the number to be separated by `_` instead of a `.`. I believe OpenPype had settings where one could choose one over the other, correct?

Like cases like: `my_image_1001_1002.exr`

kalisp replied:
Added test, works.

iLLiCiTiT commented on tests/unit/openpype/conftest.py:
```diff
@@ -0,0 +1,38 @@
+    os.environ["OPENPYPE_DATABASE_NAME"] = "avalon"
+    os.environ["AVALON_TIMEOUT"] = '3000'
+    os.environ["OPENPYPE_DEBUG"] = "3"
+    os.environ["AVALON_CONFIG"] = "pype"
```
```suggestion
```

iLLiCiTiT commented on tests/unit/openpype/conftest.py:
```diff
@@ -0,0 +1,38 @@
+
+
+if not os.environ.get("IS_TEST"):  # running tests from cmd or CI
+    os.environ["AVALON_MONGO"] = "mongodb://localhost:27017"
```
```suggestion
```

iLLiCiTiT commented on tests/unit/openpype/conftest.py:
```diff
@@ -0,0 +1,38 @@
+    os.environ["AVALON_MONGO"] = "mongodb://localhost:27017"
+    os.environ["OPENPYPE_MONGO"] = "mongodb://localhost:27017"
+    os.environ["AVALON_DB"] = "avalon"
+    os.environ["OPENPYPE_DATABASE_NAME"] = "avalon"
```
```suggestion
    os.environ["OPENPYPE_DATABASE_NAME"] = "openpype"
```

iLLiCiTiT commented on openpype/plugins/publish/validate_sequence_frames.py:
```diff
@@ -7,28 +12,51 @@ class ValidateSequenceFrames(pyblish.api.InstancePlugin):
+            return
+        for repr in representations:
+            repr_files = repr["files"]
+            if isinstance(repr_files, str):
```
Question 1: Do we want to do it for every representation? What if there is thumbnail?
Question 2: If files has string, it will 100% crash because of remainders.

tokejepsen commented:
> Previous implementation of validator failed on files with any other number in rendered file names.

Just tested latest develop with renders from Maya 2023 with this naming convention `Main001.0001.exr`, which published successfully.

Anyway of replicating the bug from Maya?

BigRoy commented:
> Just tested latest develop with renders from Maya 2023 with this naming convention Main001.0001.exr, which published successfully.
> 
> Anyway of replicating the bug from Maya?

Did you render a single frame or multiple, clique will sometimes identify the right sequence due to finding multiple frames for it. So try it out with a single frame to be sure.

kalisp commented:
Changed logic of validator, it doesn't make sense to check (missing) sequence frames for single frame. (this should handle thumbnails and reviews)

Updated tests accordingly.

tokejepsen commented:
> Did you render a single frame or multiple, clique will sometimes identify the right sequence due to finding multiple frames for it. So try it out with a single frame to be sure.

I still get a successful publish in `develop`.

@kalisp any further testing notes on how to do step 1?

kalisp commented:
> > Did you render a single frame or multiple, clique will sometimes identify the right sequence due to finding multiple frames for it. So try it out with a single frame to be sure.
> 
> I still get a successful publish in `develop`.
> 
> @kalisp any further testing notes on how to do step 1?

With `develop` Publish job on Deadline should fail for subsets like these:
![2023-03-31 14_27_21-Testing Platform PC - AnyDesk](https://user-images.githubusercontent.com/4457962/229121021-83fbc248-ee44-489e-9720-bf113e51b19e.png)

LiborBatek commented:
> > With `develop` Publish job on Deadline should fail for subsets like these:
> 
> Actually it will only for `imagesequence` family which I am not really sure how to create. I enhanced new plugin to check also "render" family, which is more sane, imho.

Basically if you succefully render out any shot from maya on Deadline and it got published to DB you will see ```Image Sequence``` asset in ```Loader``` aka OP>Load 
So any render output in DB will do for testing...speaking now of ```maya``` host

kalisp commented:
> Sorry, I'm not sure how to create such a subset through the UI. Either Maya or Nuke.

Oh, I just put `.1001` into Variant field in Nuke's New Publisher.

tokejepsen commented:
> Oh, I just put .1001 into Variant field in Nuke's New Publisher. (But if you mean how to create imagesequence family, IDK, I manipulated plugin (on develop) to trigger even on render, that I know how to do ;).)

When trying to publish through Nuke, I get this error:

```
Traceback (most recent call last):
  File "C:\Users\tokejepsen\OpenPype\.venv\lib\site-packages\pyblish\plugin.py", line 522, in __explicit_process
    runner(*args)
  File "C:\Users\tokejepsen\OpenPype\openpype\plugins\publish\collect_scene_loaded_versions.py", line 38, in process
  File "C:\Users\tokejepsen\OpenPype\openpype\hosts\nuke\api\pipeline.py", line 508, in ls
    container = parse_container(n)
  File "C:\Users\tokejepsen\OpenPype\openpype\hosts\nuke\api\pipeline.py", line 451, in parse_container
    data = read_avalon_data(node)
  File "C:\Users\tokejepsen\OpenPype\openpype\hosts\nuke\api\lib.py", line 636, in read_avalon_data
    knob_type = nuke.knob(knob.fullyQualifiedName(), type=True)
RuntimeError: Nothing is named "%s"
```

kalisp commented:
> When trying to publish through Nuke, I get this error:

I don't see how that error could have been caused by this PR.

LiborBatek commented:
Testing in ```maya 2023``` it seems I cant even create ```Render``` Instance containg number and dot.

when trying to create variant ```RenderMain.1001``` I got errors straight away...

![image](https://user-images.githubusercontent.com/112623825/230610143-afbda80a-9ca1-4faf-b5d3-c692a5a47607.png)

Which is probably ok as its more limitation than error/bug.

I can create render variant ```RenderMain1001``` so will be testing on Deadline with this one.

iLLiCiTiT commented:
I think that dot (`.`) is not allowed character in subset name? I'm surprised that it actually allowed to hit `Create`.
